File: apps/api/app/routers/commerce_suppliers.py
from __future__ import annotations

from datetime import datetime
from decimal import Decimal
from typing import Optional

from fastapi import APIRouter, Depends, HTTPException, Query
from sqlalchemy.orm import Session

from app.deps.auth import require_write_access
from app.database import get_db
from app.models.commerce_supplier import CommerceSupplier, dump_json, load_json
from app.schemas.commerce_supplier import (
    SupplierListMeta,
    SupplierListResponse,
    SupplierRead,
    SupplierResponse,
    SupplierUpdate,
)

router = APIRouter(prefix="/suppliers", tags=["commerce-suppliers"])

DEFAULT_PERFORMANCE = {
    "onTimeDeliveryPct": 0,
    "qualityRejectRatePct": 0,
    "priceVariancePct": 0,
    "avgLeadTimeDays": 0,
}


def _to_read(row: CommerceSupplier) -> SupplierRead:
    return SupplierRead(
        id=row.id,
        company_id=row.company_id,
        vendor_code=row.vendor_code,
        name=row.name,
        email=row.email,
        phone=row.phone,
        payment_terms=row.payment_terms,
        lead_time_days=row.lead_time_days,
        rating=row.rating,
        open_pos=row.open_pos,
        spend_ytd=row.spend_ytd,
        status=row.status,
        country=row.country,
        tax_id=row.tax_id,
        website=row.website,
        currency=row.currency,
        min_order_value=row.min_order_value,
        incoterms=row.incoterms,
        buyer_name=row.buyer_name,
        contacts=load_json(row.contacts_json, []),
        addresses=load_json(row.addresses_json, []),
        contracts=load_json(row.contracts_json, []),
        bills=load_json(row.bills_json, []),
        performance=load_json(row.performance_json, DEFAULT_PERFORMANCE),
        timeline=load_json(row.timeline_json, []),
        rfq_count=row.rfq_count,
        receipt_count=row.receipt_count,
        has_stock_feed=row.has_stock_feed,
        total_pos=row.total_pos,
        notes=row.notes,
        created_at=row.created_at,
        updated_at=row.updated_at,
    )


def _append_timeline(row: CommerceSupplier, title: str, description: str = "") -> None:
    timeline = load_json(row.timeline_json, [])
    timeline.insert(
        0,
        {
            "id": f"t_{int(datetime.utcnow().timestamp() * 1000)}",
            "type": "note",
            "title": title,
            "description": description,
            "at": datetime.utcnow().isoformat(),
        },
    )
    row.timeline_json = dump_json(timeline)


@router.get("", response_model=SupplierListResponse)
def list_suppliers(
    db: Session = Depends(get_db),
    search: Optional[str] = Query(default=None),
    status: Optional[str] = Query(default=None),
) -> SupplierListResponse:
    query = db.query(CommerceSupplier).order_by(CommerceSupplier.name)

    if status:
        query = query.filter(CommerceSupplier.status == status)
    if search:
        term = f"%{search.strip().lower()}%"
        query = query.filter(
            (CommerceSupplier.name.ilike(term))
            | (CommerceSupplier.vendor_code.ilike(term))
            | (CommerceSupplier.email.ilike(term))
            | (CommerceSupplier.phone.ilike(term)),
        )

    rows = query.all()
    data = [_to_read(row) for row in rows]
    total_spend = sum(row.spend_ytd for row in rows)
    return SupplierListResponse(
        data=data,
        meta=SupplierListMeta(count=len(data), total_spend_ytd=total_spend),
    )


@router.get("/{supplier_id}", response_model=SupplierResponse)
def get_supplier(supplier_id: str, db: Session = Depends(get_db)) -> SupplierResponse:
    row = db.get(CommerceSupplier, supplier_id)
    if not row:
        raise HTTPException(status_code=404, detail="Supplier not found")
    return SupplierResponse(data=_to_read(row))


@router.patch("/{supplier_id}", response_model=SupplierResponse)
def update_supplier(
    supplier_id: str,
    payload: SupplierUpdate,
    db: Session = Depends(get_db),
    _: object = Depends(require_write_access),
) -> SupplierResponse:
    row = db.get(CommerceSupplier, supplier_id)
    if not row:
        raise HTTPException(status_code=404, detail="Supplier not found")

    updates = payload.model_dump(exclude_unset=True)
    if "status" in updates and updates["status"] != row.status:
        _append_timeline(row, f"Status changed to {updates['status']}")
    for key, value in updates.items():
        setattr(row, key, value)

    db.commit()
    db.refresh(row)
    return SupplierResponse(data=_to_read(row))

File: apps/web/src/lib/api/commerce-suppliers.ts
import type {
  Supplier,
  SupplierDetail,
  SupplierStatus,
  VendorPerformance,
} from "@/lib/mock-data/suppliers";

export type ApiSupplier = {
  id: string;
  company_id: string;
  vendor_code: string;
  name: string;
  email: string;
  phone: string;
  payment_terms: string;
  lead_time_days: number;
  rating: string;
  open_pos: number;
  spend_ytd: string;
  status: string;
  country: string;
  tax_id: string | null;
  website: string | null;
  currency: string | null;
  min_order_value: string | null;
  incoterms: string | null;
  buyer_name: string | null;
  contacts: Record<string, unknown>[];
  addresses: Record<string, unknown>[];
  contracts: Record<string, unknown>[];
  bills: Record<string, unknown>[];
  performance: Record<string, unknown>;
  timeline: Record<string, unknown>[];
  rfq_count: number;
  receipt_count: number;
  has_stock_feed: boolean;
  total_pos: number;
  notes: string | null;
  created_at: string;
  updated_at: string;
};

export type ApiSupplierListResponse = {
  data: ApiSupplier[];
  meta: { count: number; total_spend_ytd: string };
};

export type ApiSupplierResponse = {
  data: ApiSupplier;
};

export type SupplierListParams = {
  search?: string;
  status?: string;
};

const DEFAULT_PERFORMANCE: VendorPerformance = {
  onTimeDeliveryPct: 0,
  qualityRejectRatePct: 0,
  priceVariancePct: 0,
  avgLeadTimeDays: 0,
};

function formatUpdatedAt(value: string): string {
  const d = new Date(value);
  if (Number.isNaN(d.getTime())) return value.slice(0, 10);
  return d.toISOString().slice(0, 10);
}

export function apiSupplierToSupplier(row: ApiSupplier): Supplier {
  return {
    id: row.id,
    vendorCode: row.vendor_code,
    name: row.name,
    email: row.email,
    phone: row.phone,
    paymentTerms: row.payment_terms,
    leadTimeDays: row.lead_time_days,
    rating: Number(row.rating),
    openPos: row.open_pos,
    spendYtd: Number(row.spend_ytd),
    status: row.status as SupplierStatus,
    country: row.country,
    updatedAt: formatUpdatedAt(row.updated_at),
    taxId: row.tax_id ?? undefined,
    website: row.website ?? undefined,
    currency: row.currency ?? undefined,
    minOrderValue: row.min_order_value != null ? Number(row.min_order_value) : undefined,
    incoterms: row.incoterms ?? undefined,
    buyerName: row.buyer_name ?? undefined,
  };
}

export function apiSupplierToSupplierDetail(row: ApiSupplier): SupplierDetail {
  const base = apiSupplierToSupplier(row);
  const perf = row.performance as Partial<VendorPerformance>;

  return {
    ...base,
    contacts: (row.contacts as SupplierDetail["contacts"]) ?? [],
    addresses: (row.addresses as SupplierDetail["addresses"]) ?? [],
    contracts: (row.contracts as SupplierDetail["contracts"]) ?? [],
    bills: (row.bills as SupplierDetail["bills"]) ?? [],
    performance: {
      onTimeDeliveryPct: Number(perf.onTimeDeliveryPct ?? DEFAULT_PERFORMANCE.onTimeDeliveryPct),
      qualityRejectRatePct: Number(
        perf.qualityRejectRatePct ?? DEFAULT_PERFORMANCE.qualityRejectRatePct,
      ),
      priceVariancePct: Number(perf.priceVariancePct ?? DEFAULT_PERFORMANCE.priceVariancePct),
      avgLeadTimeDays: Number(perf.avgLeadTimeDays ?? DEFAULT_PERFORMANCE.avgLeadTimeDays),
    },
    timeline: (row.timeline as SupplierDetail["timeline"]) ?? [],
    totalPos: row.total_pos,
    rfqCount: row.rfq_count,
    receiptCount: row.receipt_count,
    hasStockFeed: row.has_stock_feed,
  };
}

export function buildSupplierQuery(params?: SupplierListParams): string {
  if (!params) return "";
  const q = new URLSearchParams();
  if (params.search) q.set("search", params.search);
  if (params.status) q.set("status", params.status);
  const s = q.toString();
  return s ? `?${s}` : "";
}

export type UpdateSupplierInput = {
  status?: SupplierStatus;
  paymentTerms?: string;
  leadTimeDays?: number;
  buyerName?: string;
  notes?: string;
};

export function supplierUpdateToApiPayload(input: UpdateSupplierInput) {
  return {
    status: input.status,
    payment_terms: input.paymentTerms,
    lead_time_days: input.leadTimeDays,
    buyer_name: input.buyerName,
    notes: input.notes,
  };
}
